### ISSUE TITLE
Allow "render" to return DOM for type "display"

### ISSUE THREAD

**Issue by mnlipp** [DataTables/DataTables#984]
I have a table with four buttons in a cell of each row. Creating those buttons isn't as trivial as in the datatables examples, because they are shown/hidden and enabled/disabled according to row data. The easiest way to do the creation is cloning a template and applying various jquery functions to it. This results in a DOM tree that would nicely fit in the <td/>. But the render function won't except the DOM as a result, it displays "[object]". So I have to return the (outer) html of my just built dom tree, which is then parsed again for rendering.

This slows down thing unnecessarily. I tried to work around this by returning an empty string from the render function and creating the buttons in the "createdRow" callback and I had the impression that is was faster. But this approach fails if I use the responsive extension, because it renders the cell without calling the row callback.

It would be quite intuitive to be able to return a DOM tree as result of the render function being called with type "display". An alternative would be to have something like a "createdCell" callback that gets the <td/> (and the row data) as parameters and allows you to build complex content without having an intermediate textual representation.

**Comment by ImJustAskingDude** [comment]
Hey,

I have just encountered a similar problem.
In my case, I had to create a fragment and still return the outerHTML, but the simple solution to this problem should be, to just pass the cell that we're rendering to this render callback.

createdCell is great if you want to style things, but if you have both the render and createdCell, that cell is going to change to whatever you had defined for it in the render callback the second you invalidate a row.

**Comment by mateusfig** [comment]
I quite agree your idea. createdCell would solve this issue.

**Comment by ImJustAskingDude** [comment]
I'm not amazingly familiar with how the github process works, do I have to implement this solution in a datatables fork kinda thing and submit that as an improvement ? definitely willing to do that, if someone could guide me a bit, that'd be great.

**Comment by mnlipp** [comment]
Actually, the required changes for the core are already available [here](https://github.com/DataTables/DataTables/pull/985). But I admit that I haven't looked at the effects that they have on the plugins, some of which use *extremely* complicated JavaScript patterns.

**Comment by ImJustAskingDude** [comment]
This change is exactly what the OP said, but would it be also viable to pass the cell element to the render callback as a parameter ? ( possibly in the meta object ) - I know that would kind of break the current functionality, since you have to return the element.

Here's my problem:

In my setup I have a DataTable, the table has cells that contain elements that are rendered by Knockout, so in the render callback I create a document fragment, and then I use ko.renderTemplate to render to that fragment, because this is a fragment - Knockout kicks in and kicks off its binding handler. The binding handler in turn inits a given plugin, in this case select2, and select2 manipulates the DOM and does all of its changes in the fragment, now at this point I would be able to either return the fragment itself ( using the addition you just mentioned ), or I get the first element from the fragment and return outerHTML, which is what I am doing right now.

If you were to pass the empty <td> element that we are rendering to in the render callback, I would be able to ditch creating the fragment and render directly to the passed <td> element, making the code much cleaner.

This would also solve the OPs problem, although they would have to append to the passed <td> element.
This would also kind of remove the need for a createdCell, since you would be able to do the same thing in the render callback.

Now that I've written all this, I am not sure if this would serve a greater purpose or just me, so I would love your feedback on it.

P.S. Thanks mnlipp, if my idea gets any traction, I'll make sure to follow that :)

**Comment by mnlipp** [comment]
I admit that I haven't fully understood your use case, probably because I don't know "Knockout". Just two remarks:

* If you have the possibility to create (arbitrary complex) cell content in the render function, there should be little need to post-process it further in a callback.

* The existing [createdRow](https://datatables.net/reference/option/createdRow) already gets the cells as as nodes (cell DOM elements). The first parameter is the row (as node) which, of course, includes the cells.

**Comment by ImJustAskingDude** [comment]
Knockout gives you the ability to modify the DOM "off-site", you just specify some properties ( Knockout specific ) on a DOM element, and they get automatically executed when you bind the DOM element to some context. 

In my case Knockout first renders a piece of DOM, and then applies some properties kind of automatically to that.

I agree there should be no need for further post-processing, if render already lets me do all that, but it doesn't let me modify the cell element itself like createdRow or createdCell, and createdRow and createdCell only execute on creation; say you added a class to some cell, but on the next render you wanted to remove that, there is no DataTables integrated way of doing that, is there ?

While writing the above I realized that you're absolutely right, it makes no sense for DataTables to make such a change, sorry about that ;) - I still have a question though.

Let me ask you this:
Is there a way to have the render callback specified for: filter, sort, type - but somehow leave the display rendering to some other callback ?

**Comment by mnlipp** [comment]
`render` ([see here](https://datatables.net/reference/option/columns.render)) is strictly speaking not a callback. Look at the possibility to specify a function for `render` (it's on the [page](https://datatables.net/reference/option/columns.render), but there's no anchor for the section, so scroll down until you find it). The second parameter allows you to distinguish between the cases that you have mentioned.

**Comment by ImJustAskingDude** [comment]
Yup, I know about all that, the thing is, there is no way to avoid returning something for each of the types, I think, are you saying it's going to work if I just do if(type == "display") { return; } ? I am going to go and check that out right now.

**Comment by mnlipp** [comment]
@ImJustAskingDude Something is going in the wrong direction here. This feature request, as written and intended by me, is about a feature: "*Allow "render" to return DOM for type "display"*". If you want to discuss whether it is possible to return nothing, please do this in another context.

**Comment by ImJustAskingDude** [comment]
Yeah, just as a note, it's not possible to return nothing, well - thanks for the help :)
